pkg/resource/firewall/hook.go

```diff
@@ -27,3 +36,251 @@ var (
+) error {
+	if r.ko.Spec.LoggingConfiguration != nil {
+		if err := rm.syncLoggingConfiguration(ctx, r, nil); err != nil {
+			return errors.Wrapf(err, ErrSyncingPutProperty, "LoggingConfiguration")
+		}
+	}
+	return nil
+}
+
+// deleteLoggingConfig deletes logging config from a firewall. It is necessary
+// to delete logging config prior to the deletion of firewall.
+func (rm *resourceManager) deleteLoggingConfig(
+	ctx context.Context,
+	r *resource,
+) error {
+	if r.ko.Spec.LoggingConfiguration != nil {
+		if err := rm.syncLoggingConfiguration(ctx, nil, r); err != nil {
+			return errors.Wrapf(err, ErrSyncingPutProperty, "LoggingConfiguration")
+		}
+	}
+	return nil
+}
+
+// customUpdateFirewall patches each of the resource properties in the backend AWS
+// service API and returns a new resource with updated fields.
+func (rm *resourceManager) customUpdateFirewall(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (updated *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateFirewall")
+	defer exit(err)
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+
+	if delta.DifferentAt("Spec.LoggingConfiguration") {
+		if err := rm.syncLoggingConfiguration(ctx, desired, latest); err != nil {
+			return nil, errors.Wrapf(err, ErrSyncingPutProperty, "LoggingConfiguration")
+		}
+	}
+
+	return &resource{ko}, nil
+}
+
+// addPutFieldsToSpec will describe logging config and add its
+// returned values to the Firewall spec.
+func (rm *resourceManager) addLoggingConfigToSpec(
+	ctx context.Context,
+	r *resource,
+	ko *svcapitypes.Firewall,
+) (err error) {
+	getLoggingConfigurationResponse, err := rm.sdkapi.DescribeLoggingConfigurationWithContext(ctx, rm.getLoggingConfigurationPayload(r))
+	if err != nil {
+		return err
+	}
+	ko.Spec.LoggingConfiguration = rm.setResourceLoggingConfiguration(r, getLoggingConfigurationResponse)
+	return nil
+}
+
+func customPreCompare(
+	a *resource,
+	b *resource,
+) {
+	if a.ko.Spec.LoggingConfiguration == nil && b.ko.Spec.LoggingConfiguration == nil {
+		return
+	}
+
+	if a.ko.Spec.LoggingConfiguration == nil {
+		a.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	if b.ko.Spec.LoggingConfiguration == nil {
+		b.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	sort.Slice(a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+	sort.Slice(b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
```

Comment:
Are we sure that we'll have only one of each LogType at most? If not, not sure how this sort have an impact

Reply:
yes, each logtype will be unique. There are logtypes like Flow and Alert. For each of them, there will be a unique log destination